plugins/inventory/icinga.py

```diff
@@ -0,0 +1,546 @@
+            try:
+                new_d = new_d[key]
+            except IndexError:
+                self.display.vvvv(f'Structure \'{d}\' has no index \'{index}\' for sub-structure \'{new_d}\'.')
+                raise
+            except (KeyError, TypeError):
+                self.display.vvvv(f'Strucutre \'{d}\' has no key \'{key}\' for sub-structure \'{new_d}\'.')
+                raise
+
+        return new_d
+
+
+    def _get_session(self):
+        self.session         = requests.session()
+        self.session.auth    = HTTPBasicAuth(to_bytes(self.icinga_user), to_bytes(self.icinga_password))
+        self.session.headers = {
+                                 'Accept': 'application/json',
+                                 'X-HTTP-Method-Override': 'GET',
+                               }
+        self.session.verify  = self.validate_certs
+        return self.session
+
+
+    def _validate_filter(self):
+        valid_filter    = True
+        valid_keys      = [ 'name', 'group', 'zone', 'custom', 'vars' ]
+        valid_vars_keys = [ 'match', 'in', 'is' ]
+        invalid_keys    = list()
+
+        if self.filters:
+            # Check if each key is valid
+            for key in self.filters.keys():
+                if key not in valid_keys:
+                    valid_filter = False
+                    invalid_keys.append(key)
+
+            # Check validity of each 'vars' key subkey
+            if 'vars' in self.filters:
+                for sub_key in self.filters['vars']:
+                    if sub_key not in valid_vars_keys:
+                        valid_filter = False
+                        invalid_keys.append(f'vars.{sub_key}')
+
+        else:
+            valid_filter = False
+
+        if invalid_keys:
+            raise AnsibleError(f'The following keys are not valid for \'filters\': {invalid_keys}')
+
+        return valid_filter
+
+
+    def _create_filter_general(self, filter_string_base, key, values):
+        # Creates a filter for multiple list entries for a given key
+        # Handles logical concatenation and negation
+        # For each key at least on entry in the list must match their hosts according attribute
+        # The kind of filter ('match', 'in', 'is') is passed to this function
+
+        # Create filters for positive matches logically combined by 'or', and negated matches logically combined by 'and'
+        # Combine both with a logical 'and'
+        # E.g.: name must match (A or B or C) AND must match (NOT D and NOT E)
+        sub_filter          = list()
+        sub_filter_positive = list()
+        sub_filter_negative = list()
+
+        for value in values:
+            tmp_string = filter_string_base.format(value, key)
+            if value.startswith('!'):
+                tmp_string = '!{}'.format(tmp_string.replace('!', ''))
+                sub_filter_negative.append(tmp_string)
+            else:
+                sub_filter_positive.append(tmp_string)
```

Comment:
Couldn't this be done as the default without the else?

Reply:
Here the string is appended to different lists depending on whether the string starts with a '!' indicating negation. The negated and non-negated filters are concatenated differently.  
We could of course use a `continue` within the `if` block and remove the `else` if that's considered more readable.